### ISSUE TITLE
Incorrect getLogicalSizeInBytes implementation in DictionaryBlock

### ISSUE THREAD

**Issue by haozhun** [prestodb/presto#11758]
Whenever `getLogicalSizeInBytes` invokes `getRegionSizeInBytes`, it's incorrect. 

`DictionaryBlock`s can be nested in another `DictionaryBlock`. For example, a `map(varchar, varchar)` could be represented in Presto with a block like this: `Dict(Map(key=Dict(VarWidth), value=Dict(VarWidth)))`. Such a `Block` can show up in Presto today. The leaf `Block`s could be dictionary encoded by nature of ORC file encoding. The root `Block` could be dictionary encoded as it goes through a join where a probe matches multiple build entries.

In the above case, the outer `DictionaryBlock`'s `getLogicalSizeInBytes` would invoke the `MapBlock`'s `getRegionSizeInBytes`, which would in turn invoke the inner `DictionaryBlock`'s `getRegionSizeInBytes`. Since `getRegionSizeInBytes` does not return the "logical" value, it's incorrect.

In addition, there's a separate performance issue in the implementation of `getLogicalSizeInBytes`. This bug also exists in `getSizeInBytes`, and is copied into the implementation of `getLogicalSizeInBytes`. The is also incorrect in `getSizeInBytes` (although only a perf issue in `getLogicalSizeInBytes`).

To summarize,

* It's very expensive because the inner map's `getRegionSizeInBytes` will be called `m` times (`m` is the number of maps in the `MapBlock`).
* In the case of `getSizeInBytes`, it's incorrect because the inner `DictionaryBlock` entries are de-duplicated independently, leading to insufficient de-duplication.
* In the case of `getLogicalSizeInBytes`, it's incorrect because it `getRegionSizeInBytes` does not return "logical" size, which could underestimate significantly.

**Comment by dain** [comment]
I would disallow a dictionary in a dictionary and instead, rewrite the indexes so you have at most one level.

**Comment by haozhun** [comment]
The dictionaries are not adjacent in the example I gave. How do you rewrite indexes there?
